src/core/algorithms/nar/nar.cpp

```diff
@@ -0,0 +1,119 @@
+#include "nar.h"
+
+namespace model {
+std::string NAR::ToString() const {
+    std::string result;
+    result += std::to_string(qualities_.fitness);
+    result += " {";
+    size_t antecounter = 0;
+    for (auto const& [key, value] : ante_) {
+        if (antecounter > 0) {
+            result += ", ";
+        }
+        result += std::to_string(key);
+        result += ": ";
+        result += value->ToString();
+        antecounter++;
+    }
+    result += "} ===> {";
+    size_t conscounter = 0;
+    for (auto const& [key, value] : cons_) {
+        if (conscounter > 0) {
+            result += ", ";
+        }
+        result += std::to_string(key);
+        result += ": ";
+        result += value->ToString();
+        conscounter++;
+    }
+    result += "} s: ";
+    result += std::to_string(qualities_.support);
+    result += " c: ";
+    result += std::to_string(qualities_.confidence);
+    return result;
+}
+
+NARQualities CalcQualities(size_t num_rows_fit_ante, size_t num_rows_fit_ante_and_cons,
+                           size_t included_features, size_t feature_count, size_t num_rows) {
+    NARQualities result;
+    if (num_rows_fit_ante == 0) {
+        result.fitness = 0.0;
+        result.confidence = 0.0;
+        return result;
+    } else {
+        result.confidence = num_rows_fit_ante_and_cons / (double)num_rows_fit_ante;
+    }
+    result.support = num_rows_fit_ante_and_cons / (double)num_rows;
+    if (result.support == 0.0) {
+        result.fitness = 0.0;
+        result.support = 0.0;
+        return result;
+    }
+
+    double inclusion = included_features / (double)feature_count;
+    result.fitness = (result.confidence + result.support + inclusion) / 3.0;
+    return result;
+}
+
+// TODO: this function is way too big and cluttered
+void NAR::SetQualities(TypedRelation const* typed_relation) {
+    if (ante_.size() == 0 || cons_.size() == 0) {
+        qualities_.fitness = 0.0;
+        qualities_consistent_ = true;
+        return;
+    }
+
+    size_t num_rows_fit_ante = 0;
+    size_t num_rows_fit_ante_and_cons = 0;
+    for (size_t rowi = 0; rowi < typed_relation->GetNumRows(); rowi++) {
```

Comment:
extract `typed_relation->GetNumRows()` to a variable since you will be using it later